Wire-iOS/Sources/UserInterface/Helpers/RequestPasswordController.swift

```diff
@@ -0,0 +1,113 @@
+
+    enum RequestPasswordContext {
+        case removeDevice
+        case legalHold(fingerprint: Data?, hasPasswordInput: Bool)
+    }
+
+    init(context: RequestPasswordContext,
+         callback: @escaping (Result<String>) -> ()) {
+
+        self.callback = callback
+
+        let title: String
+        let message: String
+        let okTitle: String
+        let cancelTitle: String
+
+        switch context {
+        case .removeDevice:
+            title = "self.settings.account_details.remove_device.title".localized
+            message = "self.settings.account_details.remove_device.message".localized
+
+            okTitle = "general.ok".localized
+            cancelTitle = "general.cancel".localized
+
+        case .legalHold(let fingerprint, let hasPasswordInput):
+            title = "legalhold_request.alert.title".localized
+
+            let fingerprintString: String
+            if let fingerprint = fingerprint {
+                fingerprintString = (fingerprint as NSData).fingerprintString
+            } else {
+                fingerprintString = ""
+            }
+
+            var legalHoldMessage = "legalhold_request.alert.detail".localized(args: fingerprintString)
+            if hasPasswordInput {
+                legalHoldMessage += "\n"
+                legalHoldMessage += "legalhold_request.alert.detail.enter_password".localized
+            }
+            message = legalHoldMessage
+
+            okTitle = "general.skip".localized
+            cancelTitle = "general.accept".localized
+        }
+
+        alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
+
+        switch context {
+        case .removeDevice,
+             .legalHold(_, true):
+            alertController.addTextField {(textField: UITextField) -> Void in
+                textField.placeholder = "self.settings.account_details.remove_device.password".localized
+                textField.isSecureTextEntry = true
+                textField.addTarget(self,
+                                    action: #selector(RequestPasswordController.passwordTextFieldChanged(_:)),
+                                    for: .editingChanged)
+            }
+        case .legalHold(_, false):
+            break
+        }
+
+        okAction = UIAlertAction(title: okTitle, style: .default) {
+            [weak self, unowned alertController] (action: UIAlertAction) -> Void in
+            if let passwordField = alertController.textFields?[0] {
+                let password = passwordField.text ?? ""
+                self?.callback(.success(password))
+            }
+        }
+        
+        let cancelAction = UIAlertAction(title: cancelTitle, style: .cancel) {
+            [weak self, unowned alertController] (action: UIAlertAction) -> Void in
+            self?.callback(.failure(NSError(domain: "\(type(of: alertController))", code: 0, userInfo: [NSLocalizedDescriptionKey: "User cancelled input"])))
```

Comment:
Code could be `NSUserCancelledError`. I'm not sure if we need a user info either, since it's never gonna be shown.

```suggestion
            self?.callback(.failure(NSError(domain: "\(type(of: alertController))", code: NSUserCancelledError, userInfo: nil)))
```